src/artkit/model/llm/vertexai/_vertexai.py

```diff
@@ -0,0 +1,212 @@
+# -----------------------------------------------------------------------------
+# © 2024 Boston Consulting Group. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+"""
+VertexAI LLM systems.
+"""
+from __future__ import annotations
+
+import logging
+from abc import ABCMeta
+from collections.abc import Mapping
+from contextlib import AsyncExitStack
+from typing import Any, TypeVar
+
+from artkit.model.llm.history._history import ChatHistory
+from pytools.api import MissingClassMeta, appenddoc, inheritdoc, subsdoc
+
+from ...util import RateLimitException
+from ..base import ChatModelConnector
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["VertexAIChat"]
+
+try:
+    # note that vertexai is imported through the google-cloud-aiplatform package
+    import vertexai  # type: ignore
+    from google.api_core.exceptions import TooManyRequests
+    from vertexai.generative_models import (  # type: ignore
+        Content,
+        GenerativeModel,
+        Part,
+    )
+
+except ImportError:
+
+    class GenerativeModelError(
+        metaclass=MissingClassMeta, module="vertexai.generative_models"
+    ):
+        """Placeholder class for missing ``GenerativeModel`` class."""
+
+    class VertexAIError(metaclass=MissingClassMeta, module="vertexai"):
+        """Placeholder class for missing ``VertexAIError`` class."""
+
+
+__all__ = ["VertexAIChat"]
+
+#
+# Type variables
+#
+
+T_VertexAIChat = TypeVar("T_VertexAIChat", bound="VertexAIChat")
+
+#
+# Class declarations
+#
+logger = logging.getLogger(__name__)
+
+
+@inheritdoc(match="""[see superclass]""")
+class VertexAIChat(ChatModelConnector[GenerativeModel], metaclass=ABCMeta):
+    """
+    Base class for Vertex AI LLMs.
+    """
+
+    region: str | None
+    gcp_project_id: str | None
+    #: If ``False``, disable all accessible safety categories; otherwise, enable them.
+    safety: bool
+
+    @classmethod
+    def get_default_api_key_env(cls) -> str:
+        """[see superclass]"""
+        return ""
+
+    def _make_client(self) -> GenerativeModel:  # pragma: no cover
+        vertexai.init(project=self.gcp_project_id, location=self.region)
```

Comment:
the region and project id directly determines the endpoint that vertex AI calls, so if the user wants to interact with different regions or projects they should initialize a new VertexAIChat, let me know if this doesn't answer the question! @matthew-wong-bcg

Reply:
Right - but my question pertains to the use case where the following applies

```
# Initialize the first VertexAIChat
first_chat = VertexAIChat(region='us-west-1', project_id='test_project_2')
# Initialize the second VertexAIChat
second_chat = VertexAIChat(region='us-west-1', project_id='test_project_2')
```
If the `first_chat` is used anywhere, because of the way that the region and projects are used in the constructor, will calls be made to the `region` and `project_id` from the `second_chat`?

Reply:
The short answer would be no, since first_chat and second_chat are 2 different instances of vertexAI chat, all calls made using first_chat will always go to us-west-1 and test_project_1 whereas second_chat will always go to us-west-1 and test_project_2, it will be 2 different endpoints. Does this answer your question?